weavetest/abci.go

```diff
@@ -0,0 +1,137 @@
+package weavetest
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// Tester is implemented by both *testing.T and *testing.B. Use it instead of
+// the pointer type to allow notation to accept both objects.
+type Tester interface {
+	Helper()
+	Errorf(string, ...interface{})
+	Fatalf(string, ...interface{})
+	Logf(string, ...interface{})
+}
+
+// WeaveRunner provides a translation layer between an ABCI interface and a
+// weave application. It takes care of serializing messages and creating
+// blocks.
+type WeaveRunner struct {
+	chainID string
+	height  int64
+	t       Tester
+	app     abci.Application
+}
+
+// NewWeaveRunner creates a WeaveRunner instance that can be used to process
+// deliver and check transaction requests using weave API. This runner expects
+// all operations to succeed. Any error results in test failure.
+func NewWeaveRunner(t Tester, app abci.Application, chainID string) *WeaveRunner {
+	return &WeaveRunner{
+		chainID: chainID,
+		height:  0,
+		t:       t,
+		app:     app,
+	}
+}
+
+// WeaveApp is implemented by a weave application. This is the minimal
+// interface required by the WeaveRunner to be able to connect ABCI and weave
+// APIs together.
+type WeaveApp interface {
+	DeliverTx(weave.Tx) error
+	CheckTx(weave.Tx) error
+}
+
+// InitChain serialize to JSON given genesis and loads it. Loading a genesis is
+// causing a block creation.
+func (w *WeaveRunner) InitChain(genesis interface{}) {
+	raw, err := json.MarshalIndent(genesis, "", "  ")
+	if err != nil {
+		w.t.Fatalf("cannot JSON serialize genesis: %s", err)
+	}
+
+	// Load the genesis in a separate block.
+	changed := w.InBlock(func(WeaveApp) error {
```

Comment:
InitChain is not called InBlock.
Rather it is called before any block is created.

You can assert that `w.height == 0` before calling `InitChain`, but just call it directly